File: 2.dart
class Solution {
  ListNode? addTwoNumbers(ListNode? l1, ListNode? l2) {
    if (l1 == null && l2 == null) {
      return null;
    }

    int value = (l1?.val ?? 0) + (l2?.val ?? 0);
    int remainder = value ~/ 10;
    value %= 10;

    final remainNode = l1?.next ?? l2?.next;
    final node = new ListNode(value);

    if (remainNode == null) {
      if (remainder != 0) {
        node.next = ListNode(remainder);
      }
    } else {
      remainNode.val += remainder;
      node.next = addTwoNumbers(l1?.next, l2?.next);
    }

    return node;
  }
}

class ListNode {
  int val;
  ListNode? next;

  ListNode([this.val = 0, this.next]);
}

File: 2.cpp
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  static int val(ListNode *node) { return node ? node->val : 0; }

  static ListNode *next(ListNode *node) { return node ? node->next : nullptr; }

  ListNode *addTwoNumbers(ListNode *l1, ListNode *l2, int carry = 0) {
    if (!l1 && !l2 && !carry)
      return nullptr;

    auto sum = val(l1) + val(l2) + carry;
    auto *reusedNode = l1 ?: l2 ?: new ListNode;

    reusedNode->val = sum % 10;
    reusedNode->next = addTwoNumbers(next(l1), next(l2), sum / 10);
    return reusedNode;
  }
};